hkmc2/shared/src/test/mlscript-compile/Rendering.mls

```diff
@@ -35,27 +35,350 @@ fun interleave(sep)(...args) =
-      (p is Object and p.("writable")) || (p is undefined) then
-        "[function" + (if arg.name is
-          ""  then ""
-          nme then " " + nme
-        ) + "]"
-    Object then
-      if arg.constructor.name is "Object"
-      then
-        let es = Object.entries(arg)
-        fold(+)("{", ...interleave(", ")(...map(case [k, v] then k + ": " + render(v))(...es)), "}")
-      else String(arg)
-  let ts = arg.("toString") // not accessing as `arg.toString` to avoid the sanity check
-  ts is undefined then "[" + typeof(arg) + "]"
-  else ts.call(arg)
+// * Apply `f` to `xs` and build a string from non-null results using the given
+// * separator and starting/ending symbols. If there is no non-null results,
+// * return `empty` instead.
+fun mkString(xs, f, start, sep, end, empty) =
+  let
+    it = xs.(Symbol.iterator)()
+    res = it.next()
+    buffer = null
+  // Consume the iterator until we find the first non-null value.
+  while res.done is false and buffer is null do
+    if f(res.value) is
+      Str as str do set buffer = start + str
+    set res = it.next()
+  // If the buffer is still null, this means we have drained the iterator.
+  if buffer is null then empty else...
+  while res.done is false do
+    let value = f(res.value)
+    if value is Str as s do
+      set { buffer += sep, buffer += s }
+    set res = it.next()
+  set buffer += end
+  buffer
+
+// Do not wrap property names in quotes if they match the following pattern.
+let identifierPattern = new RegExp of
+  """^[\p{ID_Start}_$][\p{ID_Continue}_$]*$""", "u"
+
+fun noNeedForQuotes(id) = identifierPattern.test(id)
+
+let MIN_BREAK_LENGTH = 60
+
+let NULL_PROTO = "[Object: null prototype] "
+
+let symbolsForArray =
+  singleLineStart: "[ ", singleLineEnd: " ]"
+  multiLineStart: "[", multiLineEnd: "]"
+  singleLineSeparator: ", ", multiLineSeparator: ",", empty: "[]"
+
+let symbolsForSet(n: Int) =
+  let headline = "Set(" + n + ")"
+  (
+    singleLineStart: headline + " { ", singleLineEnd: " }"
+    multiLineStart: headline + " {", multiLineEnd: "}"
+    singleLineSeparator: ", ", multiLineSeparator: ",", empty: headline + " {}"
+  )
+
+let symbolsForObject =
+  start: "{", end: "}", entrySeparator: ",", keyValueSeparator: ": ", empty: "{}"
+
+let symbolsForMap(n: Int) =
+  let headline = "Map(" + n + ")"
+  (
+    start: headline + " {", end: "}",
+    entrySeparator: ",", keyValueSeparator: " => ", empty: headline + " {}")
+
+fun render(target, ...args) =
+  let
+    options = if args is Array and args.length > 0 then args.[0] else new Object
+    indent = if options.indent is
+      Int as value and value > 0 then " ".repeat(value)
+      true then "  ", else null
+    // If the joined string is less than or equal `breakLength`, it stays on one
+    // line; otherwise it becomes multiline and each item is prefixed with `\n`
+    // plus `indent`.
+    breakLength = if
+      options.breakLength is Int as value then Math.max(MIN_BREAK_LENGTH, value)
```

Comment:
Why would we ignore `50` and use the completely arbitrary `60` here?

I would expect a `MIN_BREAK_LENGTH` to be the smallest value that's needed for rendering to be well-defined (like `1` or something like that).